### PR TITLE
[ISSUE #15279]feat(lock): implement distributed reentrant/non-reentrant lock

### PR DESCRIPTION
**Pull request by a981008**
Phase 1 of the Nacos distributed lock system:
- Reentrant and non-reentrant lock types with owner-based identity
- Watchdog auto-renewal mechanism (30s timeout, 10s renewal interval)
- Wait queue with FIFO semantics and gRPC server-push notifications
- Raft consensus integration for strong consistency across cluster
- Connection disconnect detection with automatic lock release via Raft
- Lock expiration scanner with Raft-path release
- Defensive copy for waitQueue access and null-owner guard in scanner
- Client SDK implementing JUC Lock interface (lock/tryLock/unlock)
- 44 unit tests + 35 integration tests covering core flows and edge cases

refs #15279 

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [x] Run `mvn -B clean package apache-rat:check spotbugs:check -DskipTests` to make sure basic checks pass. Run `mvn clean install` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: api/src/main/java/com/alibaba/nacos/api/lock/LockService.java [hunk 1]
```diff
@@ -73,7 +73,16 @@ public interface LockService {
...
-    
+
+    /**
+     * Renew lock lease time (watchdog heartbeat).
+     *
+     * @param instance instance with owner set
+     * @return true if renewed successfully, false if lock not held by this owner
+     * @throws NacosException NacosException
+     */
+    Boolean renew(LockInstance instance) throws NacosException;
```
Comment (KomachiSion) [inline api/src/main/java/com/alibaba/nacos/api/lock/LockService.java:85]: Please add Since, I think this feature might be support 3.3.0

File: console/src/main/java/com/alibaba/nacos/Nacos.java [hunk 2]
```diff
@@ -46,6 +48,8 @@
 public class Nacos {
```
Comment (KomachiSion) [inline console/src/main/java/com/alibaba/nacos/Nacos.java:48]: Please use NacosBoostrap releace, Nacos in console will be remove

**Review by KomachiSion — CHANGES_REQUESTED** [review]

**Comment by KomachiSion** [comment]
Thanks for the large Phase 1 implementation. I checked the PR against #15279 and found several core lock-semantics issues that should be fixed before merge.

感谢完成这么大范围的 Phase 1 实现。我对照 #15279 检查后，发现目前还有几个会影响锁核心语义的问题，建议合并前修复。

---

1. Non-reentrant `lock()` may self-deadlock on same-thread reentry  
1. 非可重入锁同线程再次 `lock()` 可能自锁死

`NacosLock.lock()` does not check whether the current thread already holds a `NON_REENTRANT` lock before sending the acquire request. If the same owner calls `lock()` again, `NonReentrantAtomicLock#doTryLock` returns false, then the server adds the same owner into the wait queue. Since the same thread is blocked inside `lock()`, it cannot release the lock, so this can wait/retry forever.

`NacosLock.lock()` 在发送加锁请求前，没有检查当前线程是否已经持有 `NON_REENTRANT` 锁。如果同一个 owner 再次调用 `lock()`，`NonReentrantAtomicLock#doTryLock` 会返回 false，随后服务端会把同一个 owner 加入等待队列。由于同一个线程已经阻塞在 `lock()` 中，无法释放锁，因此可能无限等待/重试。

Expected behavior from the issue/design: the Java SDK should reject same-thread reentry for non-reentrant locks locally, for example by throwing `IllegalMonitorStateException`, instead of sending the request to the server.

根据 issue/design 的预期，Java SDK 应该在本地拒绝非可重入锁的同线程重入，例如抛出 `IllegalMonitorStateException`，而不是继续向服务端发送请求。

Related code:
- `client/src/main/java/com/alibaba/nacos/client/lock/NacosLock.java`
- `lock/src/main/java/com/alibaba/nacos/lock/core/reentrant/mutex/NonReentrantAtomicLock.java`
- `lock/src/main/java/com/alibaba/nacos/lock/service/impl/LockOperationServiceImpl.java`

---

2. Disconnect cleanup only releases one reentrant level  
2. 连接断开清理只释放了一层可重入计数

`releaseLocksByConnection()` calls `unLock(instance)` only once for each lock held by the disconnected connection. For `ReentrantAtomicLock`, one unlock only decrements `reentrantCount` by 1. If the client acquired the same lock multiple times before disconnecting, the lock remains owned by the dead connection.

`releaseLocksByConnection()` 对断开连接持有的每把锁只调用了一次 `unLock(instance)`。但对于 `ReentrantAtomicLock`，一次 unlock 只会让 `reentrantCount` 减 1。如果客户端断开前多次重入获取了同一把锁，锁仍然会被已经断开的连接持有。

This breaks the PR goal of automatically releasing locks on connection disconnect. The cleanup path probably needs to force-release the lock through Raft, or repeatedly release until the reentrant count reaches 0.

这会破坏 PR 中“连接断开后自动释放锁”的目标。这里可能需要通过 Raft 做 force release，或者循环释放直到 `reentrantCount` 归零。

Related code:
- `lock/src/main/java/com/alibaba/nacos/lock/service/impl/LockOperationServiceImpl.java`
- `lock/src/main/java/com/alibaba/nacos/lock/core/reentrant/mutex/ReentrantAtomicLock.java`

---

3. FIFO wait queue is not actually enforced  
3. FIFO 等待队列语义没有真正强制

On release/expire, the server polls the first waiter and sends a notification. However, after the lock becomes clear, `acquireLock()` still allows any requester to acquire immediately. A new client that was not in the wait queue can win the race before the notified waiter retries.

释放/过期时，服务端会 poll 队头等待者并发送通知。但锁变为空闲后，`acquireLock()` 仍然允许任意请求者立即获取锁。一个不在等待队列中的新客户端可能在被通知的队头等待者重试前抢到锁。

So the current implementation provides “notify first waiter”, but not true FIFO acquisition semantics. If Phase 1 wants FIFO semantics as described in #15279, the acquire path needs to respect the queue head owner when waiters exist.

因此当前实现更像是“优先通知队头”，但并没有真正保证 FIFO 获取语义。如果 Phase 1 要实现 #15279 中描述的 FIFO 语义，acquire 路径需要在存在等待者时校验队头 owner。

Related code:
- `lock/src/main/java/com/alibaba/nacos/lock/service/impl/LockOperationServiceImpl.java`
- `lock/src/main/java/com/alibaba/nacos/lock/core/reentrant/mutex/ReentrantAtomicLock.java`

---

4. Blocking lock retries can enqueue duplicate waiters  
4. 阻塞式加锁重试可能重复入队

`LockGrpcClient.lock()` retries `ACQUIRE` while waiting for notification. Each failed acquire with `waitTimeMs > 0` calls `addWaiter()` again, and `addWaiter()` does not de-duplicate by owner/connection. This can create duplicate wait queue entries for the same client, causing queue inflation and repeated notifications.

`LockGrpcClient.lock()` 在等待通知期间会反复重试 `ACQUIRE`。每次带 `waitTimeMs > 0` 的失败获取都会再次调用 `addWaiter()`，而 `addWaiter()` 没有按 owner/connection 去重。这会让同一个客户端在等待队列里出现多次，导致队列膨胀和重复通知。

Related code:
- `client/src/main/java/com/alibaba/nacos/client/lock/remote/grpc/LockGrpcClient.java`
- `lock/src/main/java/com/alibaba/nacos/lock/core/reentrant/AbstractAtomicLock.java`

---

Suggested regression tests / 建议补充回归测试：

- same-thread reentry on non-reentrant `lock()`;
  非可重入锁同线程再次调用 `lock()`；
- disconnect cleanup after `reentrantCount > 1`;
  `reentrantCount > 1` 时连接断开后的清理；
- queued waiter cannot be bypassed by a later client;
  等待队列中的客户端不能被后来的新客户端绕过；
- repeated blocking acquire does not create duplicate wait entries.
  阻塞式加锁反复重试不会创建重复等待项。

Overall, the structure looks promising, but these cases affect the core correctness guarantees of the new distributed lock feature.

整体结构是有希望的，但这些问题会影响分布式锁功能的核心正确性保证，建议修复后再合并。

**Comment by a981008** [comment]
提交此 PR 的主要目的是同步该功能的开发状态，表明相关工作正在进行中，从而减少重复实现带来的资源浪费。

The primary purpose of this PR is to communicate the current development status of this feature, indicating that work is already underway and helping to avoid duplicated implementation efforts.

**Comment by KomachiSion** [comment]
@a981008 建议将此工作拆分成多个小ISSUE和小PR， 一个大的完整PR review起来比较费劲，也容易有遗漏。

尽量拆分成小的原子性的小PR进行。

File: api/src/main/java/com/alibaba/nacos/api/lock/common/LockConstants.java [hunk 3]
```diff
@@ -23,6 +23,10 @@
  * @date 2023/8/23 15:53
  */
 public class LockConstants {
-    
+
```
Comment (KomachiSion) [inline api/src/main/java/com/alibaba/nacos/api/lock/common/LockConstants.java]: 修改缩进之后，可能spotless校验会不通过。

File: console/src/main/java/com/alibaba/nacos/Nacos.java (hunk 2 above)
Comment (KomachiSion) [reply console/src/main/java/com/alibaba/nacos/Nacos.java:48]: 这个修改的内容还是没有回滚掉。

**Review by KomachiSion — COMMENTED** [review]
Thanks for the update. Most of the previous issues have been addressed, especially:
- non-reentrant same-thread reentry is now rejected locally;
- connection cleanup now uses Raft `CLEANUP_CONNECTION` and `forceRelease`;
- wait queue entries are deduplicated by owner + connectionId.

感谢更新。之前提到的大部分问题已经处理了，尤其是：
- 非可重入锁的同线程重入现在会在客户端本地拒绝；
- 连接断开清理改成了 Raft `CLEANUP_CONNECTION` + `forceRelease`；
- 等待队列现在会按 owner + connectionId 去重。

However, I think FIFO semantics are still not fully guaranteed.

不过我认为 FIFO 语义目前仍然没有完全保证。

---

`waiterRetry` is set by the client after any local retry, not only after receiving a server notification. For example, `LockGrpcClient.lock()` sets `waiterRetry=true` after the first failed acquire and then retries every local notification poll timeout.

`waiterRetry` 是客户端在本地重试后设置的，并不只代表“收到了服务端通知后重试”。例如 `LockGrpcClient.lock()` 在第一次获取失败后就会设置 `waiterRetry=true`，随后每次本地 notification poll timeout 后都会继续重试。

On the server side, `acquireLock()` skips the “new requests must queue behind existing waiters” check when `waiterRetry=true`, and calls `tryLock()` before verifying that the retried owner is actually the queue head.

在服务端，`acquireLock()` 遇到 `waiterRetry=true` 时会跳过“有等待者时新请求必须排队”的判断，并且会在校验当前请求者是否真的是队头之前调用 `tryLock()`。

This means a non-head waiter can still acquire the lock when it happens to retry while the lock is clear.

这意味着：如果一个非队头等待者刚好在锁空闲时发起重试，它仍然可能直接获取到锁，从而破坏 FIFO。

Example timeline / 示例时间线：

1. A holds the lock.  
   A 持有锁。

2. B and C are both waiting. Queue is `[B, C]`.  
   B 和 C 都在等待，队列为 `[B, C]`。

3. A releases the lock. Server notifies B.  
   A 释放锁，服务端通知 B。

4. Before B retries, C’s local poll timeout fires and C retries with `waiterRetry=true`.  
   在 B 重试之前，C 的本地 poll timeout 触发，C 带着 `waiterRetry=true` 发起重试。

5. The lock is clear, so server allows C to acquire.  
   此时锁是空闲的，因此服务端允许 C 获取锁。

6. B is still the queue head, but FIFO has already been broken.  
   B 仍然是队头，但 FIFO 已经被破坏。

Related code / 相关代码：

- `client/src/main/java/com/alibaba/nacos/client/lock/remote/grpc/LockGrpcClient.java`
  - `waiterRetry` is set after local retry, not necessarily after notification.
  - `waiterRetry` 是本地重试后设置的，不一定代表收到了服务端通知。

- `lock/src/main/java/com/alibaba/nacos/lock/service/impl/LockOperationServiceImpl.java`
  - `acquireLock()` allows `waiterRetry` requests to call `tryLock()` before checking whether the requester is the queue head.
  - `acquireLock()` 允许 `waiterRetry` 请求在校验是否为队头之前先调用 `tryLock()`。

Suggested fix / 建议修复：

When the wait queue is non-empty, the server should allow `tryLock()` only if:
- the request is a waiter retry, and
- the queue head matches the same owner/connectionId.

当等待队列非空时，服务端只有在以下条件同时满足时才应该允许 `tryLock()`：
- 当前请求是 waiter retry；
- 队头等待者与当前请求的 owner/connectionId 匹配。

Otherwise, the request should stay waiting or update its existing wait entry, but must not acquire the lock.

否则，该请求应该继续等待，或者只更新已有等待项，但不能直接获取锁。

A stronger approach would be to include a server-issued wait token/queue position in the notification and validate it on retry, instead of trusting a client-side boolean.

更稳妥的做法是：服务端在通知中下发 wait token / queue position，并在重试时校验，而不是信任客户端传入的 boolean 标记。

Suggested regression test / 建议补充回归测试：

- Queue B and C.
  让 B 和 C 入队。
- Release the holder.
  释放当前持有者。
- Before B retries, send a retry request from C with `waiterRetry=true`.
  在 B 重试前，让 C 带 `waiterRetry=true` 发起重试。
- Assert C cannot acquire while B is still queue head.
  断言 B 仍是队头时，C 不能获取锁。

---

There is also a related cancellation concern.

另外还有一个相关的取消等待问题。

`lockInterruptibly()` can exit on `InterruptedException` without removing the server-side wait queue entry. Local `cancelWait()` only removes the client-side future; it does not remove the server-side waiter. If the interrupted waiter is queue head, later release may notify an abandoned waiter and block following waiters until the wait deadline expires.

`lockInterruptibly()` 在收到 `InterruptedException` 后可能直接退出，但不会删除服务端等待队列中的 entry。本地的 `cancelWait()` 只会移除客户端侧 future，并不会移除服务端 waiter。如果被中断的 waiter 是队头，后续释放锁时可能会通知一个已经放弃等待的 waiter，导致后面的等待者只能等到该 entry 的 wait deadline 过期后才能继续。

This path also needs either a server-side cancel/remove-waiter operation or stricter server handling to skip stale/unreachable head waiters promptly.

这里可能需要增加服务端 cancel/remove-waiter 操作，或者让服务端能够更及时地跳过 stale / unreachable 的队头等待者。

**Review by KomachiSion — DISMISSED** [review]

**Comment by KomachiSion** [comment]
Waiting for CI/IT

**Comment by KomachiSion** [comment]
reopen for cla

**Review by KomachiSion — APPROVED** [review]
